src/omega_edit/server.ts

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as vscode from 'vscode'
+import * as fs from 'fs'
+import * as unzip from 'unzip-stream'
+import * as os from 'os'
+import * as child_process from 'child_process'
+import { HttpClient } from 'typed-rest-client/HttpClient'
+import XDGAppPaths from 'xdg-app-paths'
+
+const xdgAppPaths = XDGAppPaths({ name: 'omega_edit' })
+
+class Backend {
+  constructor(readonly owner: string, readonly repo: string) {}
+}
+
+class Artifact {
+  constructor(readonly omegaEditVersion: string) {}
+
+  name = `omega-edit-scala-server-${this.omegaEditVersion}`
+  archive = `${this.name}.zip`
+  archiveUrl = (backend: Backend) =>
+    `https://github.com/${backend.owner}/${backend.repo}/releases/download/v${this.omegaEditVersion}/${this.archive}`
+
+  scriptName = os.platform().toLowerCase().startsWith('win32')
+    ? 'example-grpc-server.bat'
+    : './example-grpc-server'
+
+  getOsFolder() {
+    if (os.platform().toLowerCase().startsWith('win')) {
+      return 'windows'
+    } else if (os.platform().toLowerCase().startsWith('darwin')) {
+      return 'macos'
+    } else {
+      return 'linux'
+    }
+  }
+}
+
+// Method to get omegaEditVersion from a JSON file
+export function getOmegaEditVersion(filePath: fs.PathLike) {
+  return JSON.parse(fs.readFileSync(filePath).toString())['omegaEditVersion']
+}
+
+export async function startServer(ctx: vscode.ExtensionContext) {
+  // Get omegaEditVersion
+  const omegaEditVersion = getOmegaEditVersion(
+    ctx.asAbsolutePath('./package.json')
+  )
+  const artifact = new Artifact(omegaEditVersion)
+  const backend = new Backend('ctc-oss', 'omega-edit')
```

Comment:
I agree with this on the yarn build side that can easily be fixed but fixing it overall to be similar to what we do for the daffodil-debugger server will take some tie for me to do. Also correct if I am wrong were moving the omega-edit into Apache's hands after a while?

Reply:
Sounds good to me.

I don't know what the plans are for Omega Edit. If it does move to ASF, then we can definitely consider downloading a release at runtime. But until then, I think it needs to be packaged as part of a release so that it can be verified and so that ASF has full control of the release.

Reply:
Yeah that makes sense do you think making a separate issue to fix it would be best?

Reply:
Honestly, I'd rather it be fixed as part of this PR. I'm not that comfortable having code in our repo, even if it's unreleased, that downloads and executes a binary that could change at any point and we wouldn't know. We don't know who is building and testing this, and I don't want anyone to be surprised when an extension downloads, from their perspective, random things from the internet.

If you you think including a fix in this PR will cause issues, I'm okay with an exception and adding it later, but I would would consider that fix very high priority.

Reply:
Yeah I think it would cause issues as getting this repo to being able to fully build the omega-edit server will take a good bit of time. The easiest solution I could in this PR in a timely manner is still download the server package from the other repo but bunding it into the VSIX which I think still has the same issues you mentioned. The easiest way to get this working the way desired is getting a separate PR can addresses this issue as I think it will be a good bit to get that fixed. I would be fine tackling it as soon as this one was merged however. The main issue is going to be that the omega-edit Scala bindings are currently only published to GitHub packages which you need a GitHub token to be able to get it.

Reply:
>  download the server package from the other repo but bunding it into the VSIX which I think still has the same issues you mentioned.

I'm okay with this. It's normal to download dependencies from a server when creating a build. We do that with Daffodil for maven dependencies, for example. And nothing really verifies that those dependencies are what they should be, though we are more thorough about it when deps are updated, and again during release vote. 

But someone should have the option to verify everything that is downloaded during a build, and know that that build will never change. What I'm trying to avoid is the potential for things to change after a build is created where it can't be controlled. We should be able to create a build, sign it to say that it is safe to use, and know that anyone who uses that build will have the exact same thing.

BTW, how is OmegaEdit packaged? Could it be downloaded via maven or some other distribution channel?

Reply:
I am not aware of how to do that but I would say possibly. So the client is all npm/typescript stuff, the middleware is Scala and the backend is C++. This is all wired together via gRPC. So we create Scala dependencies that are then used by the Scala gRPC server to call the C++ shared library.

So I have a question, would it work if I store the hash of the omega-edit Scala server package in the package.json and then the TypeScript checks to see if the hash matches and if it does it will download the file otherwise it errors out saying the hash for the omega-edit server is incorrect?

Reply:
A hash helps the issue. Though I would still prefer the download happens at build time. This makes it easier to verify exactly what we actually release when we do votes, instead of having part of the release downloaded at some later point.

Reply:
So I can make the code that does the downloading only run at when either `yarn compile` or `yarn watch` is ran so it only runs when the code is being built. It will also only download the server if that specific version isn't downloaded already does that sound reasonable?

Reply:
Yeah well we will never need to download the C++ stuff. The only thing that is currently being downloaded is the Scala server so that it can extracted and ran. We use `universal:packageBin` similarly to how it is done for the daffodil debugger in this repo. So I am unaware of how we could download that ZIP in another that is why I was suggesting a script that will download the Scala server whenever a `compile` or `watch` is ran and that specific version of the server isn't already pulled.

Reply:
I see, so omega edit creats a scala server using package bin, and then VSCode ends up executing that server. Seems reasonable then. I guess it's not clear to me where the C++ comes in to play then.

Another thing to keep in mind, because this server being downloaded is not optional, all dependencies that it bundles must follow the ASF licensing requirements--everything must either be ALv2 or compatible. We'll also need to document all of those licenses as usual in the bin.LICENSE/bin.NOTICE files.

Reply:
So we created a shared library from the C++ code, which these libraries for each OS is then packaged into SBT native package that is then downloaded by the server.

For omega-edit we used a ALv2 license, repo is at https://github.com/ctc-oss/omega-edit, so we should be good there.

Reply:
@stevedlawrence Davin believes they as that is what FOSSA checks I think

Reply:
@stevedlawrence I pushed up an update for download the omega-edit server only when `yarn compile` or `yarn watch` is ran and the server version not already being downloaded

Reply:
What is FOSSA? I don't see any reference to that acronym in the vscode repo.

Reply:
So if you go to the bottom of the repo at the end of the README where it shows `License` that part is clickable and will take you to FOSSA. It does LICENSE scanning

Reply:
@stevedlawrence I added all of the scala depedencies these are the only ones packaged into the `omega-edit-scala-server-<VERSION>.zip` that had different licenses or used ALv2 and had a notice. There are no depedencies for the C++ code that get put into the shared library so the Scala are the only ones needed. These had used ALv2 without a notice:
```bash
omega-edit*
example-grpc-server.example-grpc-server-* #this is from omega-edit we need to change its name
com.github.jnr.jffi-1.3.9-native.jar
com.github.jnr.jffi-1.3.9.jar
com.github.jnr.jnr-a64asm-1.0.0.jar
com.github.jnr.jnr-ffi-2.2.11.jar
com.github.jnr.jnr-x86asm-1.0.2.jar
com.google.android.annotations-4.1.1.4.jar
com.google.code.findbugs.jsr305-3.0.2.jar
com.google.code.gson.gson-2.8.9.jar
com.google.errorprone.error_prone_annotations-2.9.0.jar
com.google.guava.failureaccess-1.0.1.jar
com.google.guava.guava-30.1.1-android.jar
com.google.guava.listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
com.google.j2objc.j2objc-annotations-1.3.jar
com.lightbend.akka.grpc.akka-grpc-runtime_2.13-2.1.3.jar
com.thesamet.scalapb.lenses_2.13-0.11.8.jar
com.thesamet.scalapb.scalapb-runtime_2.13-0.11.8.jar
com.typesafe.akka.akka-actor_2.13-2.6.9.jar
com.typesafe.akka.akka-discovery_2.13-2.6.9.jar
com.typesafe.akka.akka-protobuf-v3_2.13-2.6.9.jar
com.typesafe.akka.akka-stream_2.13-2.6.9.jar
com.typesafe.config-1.4.0.jar
com.typesafe.ssl-config-core_2.13-0.4.2.jar
```

Reply:
Crap I keep those listed in the bin.NOTICE and accidentally added them in the commit those should be removed right?

Reply:
Yeah, the info you've added to bin.NOTICE isn't quite right. It should only include the content from the NOTICE file of the dependency if they have one. You don't need license/homepage information.

And you don't need to include deps if they don't have a NOTICE file. Though, I think maybe we started doing that to make it easier to verify that all deps are accounted for? @tuxji, I think you switch things to that? Is that the case or am I making that up?

Reply:
Yeah that is exactly what I did for the NOTICE's. Anything added to the bin.NOTICE was used in that repos NOTICE. I just removed those extras because it looked funny but can add them back in